src/core/controls/PointerLockControls.tsx

```diff
@@ -0,0 +1,161 @@
+import {
+  useEffect,
+  useCallback,
+  useState,
+  useMemo,
+  MutableRefObject,
+  useRef,
+} from "react";
+import { useFrame, useThree } from "react-three-fiber";
+import { Euler } from "three";
+import { useEnvironment } from "../utils/hooks";
+import * as THREE from "three";
+
+type Props = {
+  position: MutableRefObject<THREE.Vector3>;
+};
+
+const MIN_POLAR_ANGLE = 0; // radians
+const MAX_POLAR_ANGLE = Math.PI; // radians
+const SENSITIVITY = 0.8;
+const PI_2 = Math.PI / 2;
+
+/**
+ * PointerLockCamera is a react port of PointerLockControls.js from THREE,
+ * with some changes. Some parameters are listed above
+ *
+ * https://github.com/mrdoob/three.js/blob/master/examples/jsm/controls/PointerLockControls.js
+ *
+ * @param props = { onUnlock: function to run when the pointer lock controls are unlocked }
+ * @constructor
+ */
+const PointerLockCamera = (props: Props) => {
+  const { position } = props;
+
+  const { camera, gl } = useThree();
+  const { domElement } = gl;
+  const { paused, setPaused, addEvent } = useEnvironment();
+
+  const euler = useMemo(() => new Euler(0, 0, 0, "YXZ"), []);
```

Comment:
```suggestion
  const { current: euler } = useRef(new Euler(0, 0, 0, "YXZ"));
```
just a slightly better way to handle referential equality in this case